common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -253,6 +255,16 @@ public int numBytes() {
    * Returns the number of code points in it.
    */
   public int numChars() {
+    if (numChars == 0) numChars = getNumChars();
```

Comment:
This looks like an unrelated change (to cache numChars). Can we do it with a separated PR? We should also add proper protection for multi-thread.

Reply:
agreed, removing this change from the current PR

Reply:
opened a new PR for this: https://github.com/apache/spark/pull/47142